### DISCUSSION TITLE
Code Generation from nodeset.xml

### DISCUSSION THREAD

**Discussion by arnoweiss** [eclipse-milo/milo#877] [Ideas]
Is there a code generation tool that processes nodeset.xml (for example built in an Editor) and builds Server-SDK-based code? This could significantly accelerate the prototyping process for Milo-based servers and already exists for other OPC UA stacks.

Otherwise, we'll try at building it inhouse.

**Comment by kevinherron** [comment]
I am building code generation tools but I don't plan to have them available in any capacity until Milo version 2.0, supporting OPC UA 1.04, is released. 

If you need them now and for the current version then you'll have to build them in house.

**Reply by kevinherron** [reply]
@alw-iwu FWIW, I do have this available: https://github.com/digitalpetri/opc-ua-nodeset-parser

Some of it may be useful to you. My current prototype code generators use a slightly updated version of it, but in the rewrite I'm hoping will become the production code generators I have taken a different approach where I do not use an intermediate representation (everything in [this package](https://github.com/digitalpetri/opc-ua-nodeset-parser/tree/master/src/main/java/com/digitalpetri/opcua/nodeset/attributes)), so this repo may end up archived at some point.

**Comment by johannwesely** [comment]
Hi is there something new about reading an xml file for auto generated Server address space with milo ?

**Reply by francisjjp** [reply]
I have seen that for python (opcua. link: https://github.com/FreeOpcUa/python-opcua/blob/master/examples/server-import-robot-nodeset.py)
and C (open62541. link: https://github.com/open62541/open62541-nodeset-loader), seems, there are some kind of libraries to run in runtime a nodeset.xml and import complete OPC model in server without compilling.


Is there something like that for OPC Eclipse milo server in roadmap?

**Reply by kevinherron** [reply]
@francisjjp I do hope to support this some time after 1.0 is released later this year, but there's nothing more definitive then that.

**Reply by francy328** [reply]
I thinks still nothing about this issue?

**Reply by kevinherron** [reply]
Nothing has changed, still hope to support this some time after 1.0.

**Reply by kevinherron** [reply]
Not release ready, but the beginnings of this are at https://github.com/digitalpetri/uanodeset

Used by [DataTypeTestNamespace](https://github.com/digitalpetri/opc-ua-demo-server/blob/1.0/src/main/java/org/eclipse/milo/opcua/test/DataTypeTestNamespace.java) in the 1.0 demo server.
